### PR TITLE
Do not hold compilation monitor in postCompilationTasks on the server

### PR DESCRIPTION
**Pull request by dmitry-ten**
As described in #11831, we do not want to hold the compilation
queue monitor while sending the last message with a compiled method
back to the client. We do not actually need the monitor on the server in
`postCompilationTasks` so the acquisition is delayed until
the end of that method.

### CODE REVIEW

**Comment by dmitry-ten** [comment]
@mpirvu fyi.
I will also run the experiment with many clients connected to the same server and check if `compQueueMonitor` contestion is any better with this change.

**Comment by dmitry-ten** [comment]
Just pushed a small fix for build breaking.

File: runtime/compiler/control/JITServerCompilationThread.cpp [hunk 1]
```diff
@@ -167,6 +167,19 @@ outOfProcessCompilationEnd(
...
+   J9VMThread *vmThread = compInfoPT->getCompilationThread();
+   TR::CompilationInfo *compInfo = compInfoPT->getCompilationInfo();
+   compInfo->debugPrint(vmThread, "\tacquiring compilation monitor\n");
+   compInfo->acquireCompMonitor(vmThread);
+   compInfo->debugPrint(vmThread, "+CM\n");
+   // Acquire the queue slot monitor now
+   //
+   compInfo->debugPrint(vmThread, "\tre-acquiring queue-slot monitor\n");
+   entry->acquireSlotMonitor(vmThread);
+   compInfo->debugPrint(vmThread, "+AM-", entry);
```
Comment (mpirvu) [inline runtime/compiler/control/JITServerCompilationThread.cpp:182]: I am favoring moving this code at the end of postCompilationTasks() to minimize the time we hold the compilation monitor but also to keep the lock acquisition at the same level (for non-jitserver we acquire the compilation monitor in postCompilationTasks, so we should do it i this method for jitserver as well)

**Comment by dmitry-ten** [comment]
Updated.

**Comment by mpirvu** [comment]
I wonder what happens if we have an exception during `postCompilationtasks()`. It is possible because we allocate memory and because we send the compilation result. Then we enter again that problematic catch block where we put a fatal assert that compiler object must be NULL.

**Comment by dmitry-ten** [comment]
`ServerStream::finishCompilation` has a try/catch block inside, so it won't throw. 
If memory allocation exception is a concern, then the assert in the other PR can be just removed.

**Comment by mpirvu** [comment]
Is this still happening:
```
   if (_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
../compiler/control/CompilationThread.cpp:7735:69: error: use of undeclared identifier 'JITServer'
   if (_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
                                                                    ^
clang++-3.8 -std=c++0x -w -fsyntax-only -ferror-limit=0  -Xclang -load -Xclang /home/runner/work/openj9/openj9/runtime/omr/tools/compiler/OMRChecker/OMRChecker.so -Xclang -add-plugin -Xclang omr-checker  -DBITVECTOR_BIT_NUMBERING_MSB -DJ9_PROJECT_SPECIFIC -DTR_HOST_X86 -DTR_HOST_64BIT -DBITVECTOR_64BIT -DLINUX -DTR_TARGET_X86 -DTR_TARGET_64BIT -DSUPPORTS_THREAD_LOCAL -D_LONG_LONG -DJ9HAMMER -I'/home/runner/work/openj9/openj9/runtime/omr/include_core' -I'/home/runner/work/openj9/openj9/runtime/omr/thread' -I'/home/runner/work/openj9/openj9/runtime/omr/gc/include' -I'../compiler/x/amd64' -I'../compiler/x' -I'../compiler' -I'../omr/compiler/x/amd64' -I'../omr/compiler/x' -I'../omr/compiler' -I'../omr' -I'..' -I'/home/runner/work/openj9/openj9/runtime/codert_vm' -I'/home/runner/work/openj9/openj9/runtime/gc_include' -I'/home/runner/work/openj9/openj9/runtime/include' -I'/home/runner/work/openj9/openj9/runtime/gc_glue_java' -I'/home/runner/work/openj9/openj9/runtime/jit_vm' -I'/home/runner/work/openj9/openj9/runtime/nls' -I'/home/runner/work/openj9/openj9/runtime/oti' -I'/home/runner/work/openj9/openj9/runtime/util' -I'/home/runner/work/openj9/openj9/build/runtime' -I'/home/runner/work/openj9/openj9/build/runtime/omr' -I'/home/runner/work/openj9/openj9/build/runtime/nls' -I'/home/runner/work/openj9/openj9/build/runtime/include' -I'/home/runner/work/openj9/openj9/build/runtime/compiler' -o ../compiler/control/JitDump.cpp.linted -c ../compiler/control/JitDump.cpp
2 errors generated.
make: *** [../compiler/control/CompilationThread.cpp.linted] Error 1
linter.mk:160: recipe for target '../compiler/control/CompilationThread.cpp.linted' failed
make: *** Waiting for unfinished jobs....
Error: Process completed with exit code 2.
```

**Comment by mpirvu** [comment]
Jenkins compile xlinux jdk11

File: runtime/compiler/control/CompilationThread.cpp [hunk 2]
```diff
@@ -7724,6 +7732,20 @@ TR::CompilationInfoPerThreadBase::postCompilationTasks(J9VMThread * vmThread,
    _vm = NULL;
    _addToJProfilingQueue = false;
 
+   if (_compInfo.getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
```
Comment (mpirvu) [inline runtime/compiler/control/CompilationThread.cpp:7735]: We need the `#if defined(J9VM_OPT_JITSERVER)` here. This is why the linter job fails

**Review by mpirvu — CHANGES_REQUESTED** [review]

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

**Pull request merged**